Refactor ChangePassword

madmax330 opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/Expensify/issues/211242

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.


<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps

- Log into an account on NewDot
- Go to Profile > Security > Change Password
- Enter a wrong password and a new password
- Make sure you see the error displayed and that the api response contains the onyxData

```
{"onyxData":[{"onyxMethod":"merge","key":"account","value":{"error":"Your password changes could not be saved because the old password entered is incorrect."}}],"jsonCode":200,"requestID":"mV4I24"}
```

- Enter the correct password and a new password and everything should work.

- [x] Verify that no errors appear in the JS console

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->

madmax330 committed:
Refactor ChangePassword

techievivek reviewed (commented):
Looks good so far. Is there anything else remaining to be updated with this other than updating the PR checklist and adding tests and videos/screenshots?

madmax330 committed:
use new api

madmax330 committed:
Merge branch 'main' into maxence-rcpc
update branch

madmax330 committed:
add import

madmax330 committed:
lint fix

madmax330 committed:
small updates

madmax330 commented:
PR is ready for review, should be reviewed and tested alongside: https://github.com/Expensify/Web-Expensify/pull/33885

pulling in @iwiznia since you said you had time for refactor reviews 😄

iwiznia commented on src/libs/actions/User.js:
```diff
@@ -46,25 +47,34 @@ Onyx.connect({
-            Onyx.merge(ONYXKEYS.ACCOUNT, {loading: false});
-        });
+function changePassword(oldPassword, password) {
+    const optimisticData = [
```
NAB: I don't think this variables are needed or make things much clearer

iwiznia commented on src/pages/settings/PasswordPage.js:
```diff
@@ -133,7 +133,7 @@ class PasswordPage extends Component {
         if (!this.validate()) {
             return;
         }
-        User.changePasswordAndNavigate(this.state.currentPassword, this.state.newPassword);
```
So, this never really navigated?

madmax330 replied:
Apparently not 🤷🏽

mountiny replied:
Yeah, seems like the original function never had any navigation in it 😄

madmax330 committed:
pass values directly

mountiny reviewed (commented):
Came over out of curiosity, LGTM. It would great to leave a video of the testing steps succeeding with the Web-E branch.

techievivek reviewed (commented):
Changes seem to work fine. Can we add a video testing this new refactor? 

Also While testing this, I observed something interesting(not sure if it is a bug) but once I changed the password (and got a confirmation) I clicked on the back arrow and tried changing the password again and it shows `Session expired error` and this is nowhere visible on the UI just in the response data and spinner is in running state. But I didn't have to log in again I refreshed the page and everything worked fine.

<img width="1618" alt="Screenshot 2022-06-07 at 5 13 59 PM" src="https://user-images.githubusercontent.com/35863227/172373554-bdcc8197-f2a1-45c7-8d87-dff79cdbcfbd.png">

madmax330 committed:
Merge branch 'main' into maxence-rcpc
update branch

madmax330 committed:
Use new command

madmax330 commented:
Refactored to use the new command. Here is the video:

https://user-images.githubusercontent.com/7016692/172562142-c38e644d-9a35-4508-b860-073845ce9b3d.mov


> Also While testing this, I observed something interesting(not sure if it is a bug) but once I changed the password (and got a confirmation) I clicked on the back arrow and tried changing the password again and it shows Session expired error and this is nowhere visible on the UI just in the response data and spinner is in running state. But I didn't have to log in again I refreshed the page and everything worked fine.

Hmmm, that's interesting, although I don't think it's related to this PR.

mountiny reviewed (commented):
> Also While testing this, I observed something interesting(not sure if it is a bug) but once I changed the password (and got a confirmation) I clicked on the back arrow and tried changing the password again and it shows Session expired error and this is nowhere visible on the UI just in the response data and spinner is in running state. But I didn't have to log in again I refreshed the page and everything worked fine.

@techievivek would you be able to see if the same reproduction steps do this on the main too or it does not happen on main branch? To be sure it is related to this PR

techievivek commented:
While testing something on NewDot observed something which I think this PR is missing. Once when we are done with changing the password when we try changing the password **again** it does give us `Session expired error` but then it makes a subsequent call to `Authenticate` command and then recalls the `ChangePassword` command which is not the same for this PR as it keeps calling `ChangePassword` which returns `Session expired error`.(All this is done on main branch)
<img width="1688" alt="Screenshot 2022-06-12 at 8 43 01 PM" src="https://user-images.githubusercontent.com/35863227/173240069-3f482f13-38cd-4da9-aa22-36f13d2eee43.png">

madmax330 commented:
Hmm the only explanation I can think of is b/c we're using API.write() instead of the deprecated API.

madmax330 commented:
Either way, I'm waiting on the web PR to be complete then I'll test this again and try to figure it out.

madmax330 committed:
Merge branch 'main' into maxence-rcpc
update branch

marcaaron commented:
One idea is to just force all requests in the write queue to have a `shouldRetry` true if they are called via `API.write()`

https://github.com/Expensify/App/blob/07fabcf6c2b033faedb5215e2c9ddcc195fe98fc/src/libs/Network/index.js#L38-L39

Looks like these params are no longer included, but the reauthentication middleware still looks at them.

madmax330 committed:
Force shouldrety and cancel for deprecated api

madmax330 committed:
Fix infinite loop

madmax330 committed:
Merge branch 'main' into maxence-rcpc

update branch

madmax330 commented:
Alright, updated and retested, everything should be fine now

marcaaron commented on src/libs/API.js:
```diff
@@ -19,7 +19,14 @@ function write(command, apiCommandParameters = {}, onyxData = {}) {
+            ...data,
+
+            // We add this here because this was passed in the old api requests and some middlewares still expect this.
+            // This should be removed once we are no longer using deprecatedAPI
```
Let's create a follow up issue to do this vs. leaving this "todo" in the code.

marcaaron commented on src/libs/API.js:
```diff
@@ -19,7 +19,14 @@ function write(command, apiCommandParameters = {}, onyxData = {}) {
+        data: {
+            ...data,
+
+            // We add this here because this was passed in the old api requests and some middlewares still expect this.
```
NAB, this comment is not necessary. Whoever wants to remove this can just look to see if it's being used and why.

marcaaron reviewed (commented):
When testing I am seeing multiple API calls and calls to `UpdatePassword` and a call to `Authenticate` (even though the authToken is up to date). I might need to update Auth or something though or maybe it is expected that a successful call to `UpdatePassword` returns this response? Unsure...

marcaaron commented:
https://user-images.githubusercontent.com/32969087/174670367-41401540-ee36-4d38-b629-cf0bef9f2435.mp4

madmax330 committed:
Add issue + remove comment

madmax330 commented:
Yeah @marcaaron what you posted is the expected behavior. 
The first call to `UpdatePassword` returns 407 (unauthenticated) so it authenticates, then calls `UpdatePassword` again. 
Same as prod.

techievivek commented:
Yep, that is expected behaviour.

techievivek reviewed (approved):
Looks good to me now. Thanks

marcaaron commented:
Ok I was confused but I think what happens is....

- Call `UpdatePassword` once (returns 200 but your authToken gets invalidated)
- Call `UpdatePassword` a second time and you will have to first reauthenticate and then call `UpdatePassword` again

marcaaron commented:
Linked Web-E PR is on production so gonna merge this one 🚀

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/2575330069) to staging by @marcaaron in version: 1.1.79-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/2633881559) to production by @roryabraham in version: 1.1.79-17 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
